framework/src/headers/ip/v6/icmp/router_advertisement.rs

```diff
@@ -0,0 +1,339 @@
+        u32::from_be(self.reachable_time)
+    }
+
+    #[inline]
+    pub fn retrans_timer(&self) -> u32 {
+        u32::from_be(self.retrans_timer)
+    }
+}
+
+impl<T> IPv6Optionable for Icmpv6RouterAdvertisement<T>
+where
+    T: Ipv6VarHeader,
+{
+    fn parse_options(&self, payload_len: u16) -> HashMap<Icmpv6OptionType, Icmpv6Option> {
+        let mut options_map = HashMap::new();
+        unsafe {
+            let self_as_u8 = (self as *const Self) as *const u8;
+            let mut payload_offset = self.offset(); //track to make sure we don't go beyond v6 payload size
+
+            while payload_offset < (payload_len as usize) {
+                // start at beginning of the options
+                // the second byte is the header length in 8-octet unit excluding the first 8 octets
+                let seek_to = self_as_u8.offset(payload_offset as isize);
+
+                //lets get the first two indices which should be option_type and option_length fields
+                let option_meta = slice::from_raw_parts(seek_to, 2);
+
+                //Parse the option type field first then the option_length
+                let option_type: Option<Icmpv6OptionType> = FromPrimitive::from_u8(option_meta[0]);
+                let option_length = option_meta[1];
+                let option_length_octets = option_length * 8;
+
+                //Make sure the option is valid defined by the RFC. Note: a router can insert its own
+                //options defined outside of the spec.
+                if option_type.is_some() {
+                    //option_value_length is always total number of octets - 2(option type and length are always 2)
+                    let option_value_length = (option_length_octets - 2) as isize;
+
+                    let value_seek_to = self_as_u8.offset((payload_offset + 2) as isize);
+                    let option_value =
+                        slice::from_raw_parts(value_seek_to, option_value_length as usize);
+                    let option_type_enum = option_type.unwrap();
+                    match option_type_enum {
+                        Icmpv6OptionType::SourceLinkLayerAddress => options_map.insert(
+                            option_type_enum,
+                            Icmpv6Option::SourceLinkLayerAddress(MacAddress::new(
+                                option_value[0],
+                                option_value[1],
+                                option_value[2],
+                                option_value[3],
+                                option_value[4],
+                                option_value[5],
+                            )),
+                        ),
+                        _ => None,
+                    };
+                }
+                else {
+                   //log loudly here that option type does not exist
+                }
+
+                payload_offset += (option_length_octets) as usize;
+            }
+        }
+        options_map
+    }
+}
```

Comment:
@scrollins this is what I meant with the option type and pattern matching...

```rust
impl<T> IPv6Optionable for Icmpv6RouterAdvertisement<T>
where
    T: Ipv6VarHeader,
{
    fn parse_options(&self, payload_len: u16) -> HashMap<Icmpv6OptionType, Icmpv6Option> {
        let mut options_map = HashMap::new();
        unsafe {
            let self_as_u8 = (self as *const Self) as *const u8;
            let mut payload_offset = self.offset(); //track to make sure we don't go beyond v6 payload size

            while payload_offset < (payload_len as usize) {
                // start at beginning of the options
                // the second byte is the header length in 8-octet unit excluding the first 8 octets
                let seek_to = self_as_u8.offset(payload_offset as isize);

                //lets get the first two indices which should be option_type and option_length fields
                let option_meta = slice::from_raw_parts(seek_to, 2);

                //Parse the option type field first then the option_length
                let option_type: Option<Icmpv6OptionType> = FromPrimitive::from_u8(option_meta[0]);
                let option_length = option_meta[1];
                let option_length_octets = option_length * 8;

                //option_value_length is always total number of octets - 2(option type and length are always 2)
                let option_value_length = (option_length_octets - 2) as isize;
                let value_seek_to = self_as_u8.offset((payload_offset + 2) as isize);
                let option_value =
                    slice::from_raw_parts(value_seek_to, option_value_length as usize);

                //Make sure the option is valid defined by the RFC. Note: a router can insert its own
                //options defined outside of the spec.
                match option_type {
                    // we only care about matching on source link layer address
                    Some(Icmpv6OptionType::SourceLinkLayerAddress) => {
                        options_map.insert(
                            Icmpv6OptionType::SourceLinkLayerAddress,
                            Icmpv6Option::SourceLinkLayerAddress(
                                MacAddress::new(
                                    option_value[0],
                                    option_value[1],
                                    option_value[2],
                                    option_value[3],
                                    option_value[4],
                                    option_value[5],
                                )
                            )
                        );
                    },
                    Some(other) => {
                        // log maybe?
                    },
                    None => {
                        // log loudly as we cannot handle this option type
                    }
                }

                payload_offset += (option_length_octets) as usize;
            }
        }
        options_map
    }
}
```

Reply:
One other observation is, it would seem that the option parsing is independent of the NDP Message type.  For example, source link layer address can appear on different messages.

I would think that we should be able to pull the parsing out to separate functions (perhaps living on the `impl NDPOption` that would do what is happening here

Namely, given an option type, option length, and pointer to whatever buffer we are reading, do the right thing and read the bytes into the appropriate `Option` struct

Reply:
yeah that is true. it is independent of message type. i thought putting on a message type would probably create dup code but its probably best to move it to the NDPOption

Reply:
@scrollins the thing that is new to me here is  how we read the data in.  In my head conceptually I have a byte array under the covers (or byte stream).  We use offsets to point to different locations in the byte array.

So, the solution in my head becomes simple.  Provide a function like the following (using Scala instead since I am more comfortable there).  Also, this uses recursion but a `while` loop is fine too.

```scala
def parseOptions(bytes: Array[Byte], idx: Int, acc: Map[NDPOptionType, NDPOption]): Map[NDPOptionType, NDPOption] = {
  if (idx >= bytes.length) {
    acc
  } else {
    // recursively call to add the next option
    parseOption(bytes, idx) match {
      case (newIdx, Some((otype, oval)) => 
        // add the new option to the map and recursive to ourselves using the new offset index
        parseOptions(bytes, newIdx, acc + (otype -> oval))
      case (newIdx, None) =>
        // cannot parse this guy, just recurse with the new index and current state
        parseOptions(bytes, newIdx, acc)
      }
  }
}

// returns a tuple with the new index, and an option of an option in case we cannot parse
def parseOption(bytes: Array[Byte], offset: Int): (Int, Option[(NDPOptionType, NDPOption)]) = {
  if (offset >= bytes.length) {
    (offset, None)
  } else {
    val typ = bytes[offset]
    val len = bytes[offset + 1]

    // Note: I do not have this but it is simple to surmise what this would do if it existed
    val optionType = NDPOptionType.fromByte(bytes[idx])

    // assume this is an option and may not be there Option[NDPOption]
    // also, I don't have this, but the expectation is that this function would return an Option if it could not parse
    val option = NDPOption.parse(bytes, optionType, offset+2, len)

    // return the new offset, and the optional NDPOption
    val newIndex = offset + 2 + len
    (newIndex, option.map(opt => (optionType, opt))
}
```

Reply:
@scrollins note: on the above, I kinda pseudo coded it.  I am not all that familiar with Rust yet, and certainly not familiar with the underlying mechanism for pulling bytes off the packet

Anyway, it shows how we could write a function like `impl NDPOption { fn parse(xxx) -> yyy }` for doing the actual parsing.  In there we could do a pattern match on the option type passed in.

The `NDPOptionType.from` I think we already have.

Reply:
The one thing we do lose is the `parseOptions` can theoretically return _any_ option, even if it is not valid for the current NDP Message Type.

Could optionally add a `pub possible_options: Set<NDPOPtionType>` on the `NDPMessage` itself (could be an empty set), and pass that in as well to ensure that only valid options are parsed for each concrete NDP Message Type.  Or, we could implement `parse_options` for each NDP Message type, and call the helper and pass in the "known option types" as well.  Either way, the goal would be to pass in the allowable OptionTypes for the current NDP Message Type

Reply:
@scrollins finally (sorry for the essay I am writing here), probably makes sense to wrap up this PR, and refactor as we go for the other NDP Message types before settling on a better pattern.  tl;dr; Ideas above but probably premature